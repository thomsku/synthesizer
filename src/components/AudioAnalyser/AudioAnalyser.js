import React from "react";
import AudioVisualiser from "../AudioVisualiser/AudioVisualiser";

class AudioAnalyser extends React.Component {
  constructor(props) {
    super(props);

    this.state = {audioData: new Uint8Array(0)};
  }
  render() {
    return (
      <div>
        <AudioVisualiser audioData={this.state.audioData}/>
      </div>
    );
  }
}

export default AudioAnalyser;
