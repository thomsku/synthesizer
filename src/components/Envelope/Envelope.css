.graph {
  height: 10em;
  width: 25em;
  margin-top: 0.5em;
}

.slider {
  padding: 22px 0px;
}