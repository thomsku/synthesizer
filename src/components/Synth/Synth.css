.parameters {
  display: flex;
  justify-content: space-evenly;
}

.radio, .radiolabel {
  color: white !important;
}