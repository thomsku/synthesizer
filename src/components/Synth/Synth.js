import React from "react";

// This is tuned to just intonation
const keyToFrequency = {
  a: 261.63,
  s: 294.33,
  d: 327.03,
  f: 392.44,
  g: 436.05,
  h: 523.25,
  j: 588.66,
  k: 654.06,
  l: 697.66,
};


class Synth extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      activeOscillators: {},
    };

    this.handleKeyDown = this.handleKeyDown.bind(this);
    this.handleKeyUp = this.handleKeyUp.bind(this);
  }

  componentDidMount() {
    this.audioContext = new (window.AudioContext || window.webkitAudioContext)();
    this.gainNode = this.audioContext.createGain();
    this.gainNode.connect(this.audioContext.destination);

    document.addEventListener("keydown", this.handleKeyDown);
    document.addEventListener("keyup", this.handleKeyUp);
  }

  handleKeyDown(event) {
    const key = event.key;
    if (!this.state.activeOscillators[key] && keyToFrequency[key]) {
      const oscillator = this.audioContext.createOscillator();
      oscillator.frequency.setValueAtTime(keyToFrequency[key], this.audioContext.currentTime);
      this.setState((prevState) => ({
        activeOscillators: {
          ...prevState.activeOscillators,
          [key]: oscillator,
        },
      }));
      this.state.activeOscillators[key].connect(this.gainNode);
      this.state.activeOscillators[key].start();
    }
  }

  handleKeyUp(event) {
    const key = event.key;
    if (this.state.activeOscillators[key]) {
      this.state.activeOscillators[key].stop();
      delete this.state.activeOscillators[key];
    }
  }

  render() {
    return (
      <div className="synthDiv">

      </div>
    );
  }
}

export default Synth;
