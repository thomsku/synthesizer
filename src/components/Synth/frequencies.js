// This is tuned to just intonation
export const justKeyToFrequency = {
  a: 261.63, // C
  s: 294.33, // D
  d: 327.03, // E
  f: 392.44, // G
  g: 436.05, // A
  h: 523.25, // C
  j: 588.66, // D
  k: 654.06, // E
  l: 697.66, // F
};

// This is tuned to equal temperament
export const equalKeyToFrequency = {
  a: 261.63, // C
  s: 293.66, // D
  d: 329.63, // E
  f: 392.00, // G
  g: 440.00, // A
  h: 523.25, // C
  j: 587.32, // D
  k: 659.26, // E
  l: 698.46, // F
};
