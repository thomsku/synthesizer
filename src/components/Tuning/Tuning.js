import React from "react";
import Radio from "@material-ui/core/Radio";
import RadioGroup from "@material-ui/core/RadioGroup";
import FormControlLabel from "@material-ui/core/FormControlLabel";
import FormLabel from "@material-ui/core/FormLabel";
import PropTypes from "prop-types";
import "./Tuning.css";

const Tuning = ({
  setTuning,
  tuning,
}) => {
  return (
    <div className="radiobuttons">
      <FormLabel className="parametertitle" component="legend">Tuning</FormLabel>
      <RadioGroup
        aria-label="Tuning"
        name="tuning"
        className="radiogroup"
        value={tuning}
        onChange={setTuning}
      >
        <FormControlLabel className="radiolabel" value="just" control={<Radio className="radio"/>} label="just intonation" />
        <FormControlLabel className="radiolabel" value="equal" control={<Radio className="radio"/>} label="equal temperament" />

      </RadioGroup>
    </div>
  );
};

Tuning.propTypes = {
  setTuning: PropTypes.func.isRequired,
  tuning: PropTypes.string.isRequired,
};

export default Tuning;
