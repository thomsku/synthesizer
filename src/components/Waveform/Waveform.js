import React from "react";

class Waveform extends React.Component {
  constructor(props) {
    super(props);

    this.state = {};
  }
  render() {
    return (
      <div></div>
    );
  }
}

export default Waveform;
