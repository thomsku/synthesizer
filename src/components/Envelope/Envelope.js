import React from "react";
import EnvelopeGraph from "adsr-envelope-graph";
import Slider from "@material-ui/lab/Slider";
import Typography from "@material-ui/core/Typography";
import FormLabel from "@material-ui/core/FormLabel";
import {withStyles} from "@material-ui/core/styles";
import {fade} from "@material-ui/core/styles/colorManipulator";
import PropTypes from "prop-types";
import "./Envelope.css";

const StyledSlider = withStyles({
  thumb: {
    height: 24,
    width: 24,
    backgroundColor: "#fff",
    "&$focused, &:hover": {
      boxShadow: `0px 0px 0px ${8}px ${fade("#de235b", 0.16)}`,
    },
    "&$activated": {
      boxShadow: `0px 0px 0px ${8 * 1.5}px ${fade("#de235b", 0.16)}`,
    },
    "&$jumped": {
      boxShadow: `0px 0px 0px ${8 * 1.5}px ${fade("#de235b", 0.16)}`,
    },
  },
  track: {
    backgroundColor: "#fff",
    height: 8,
  },
  trackAfter: {
    backgroundColor: "#fff",
  },
  focused: {},
  activated: {},
  jumped: {},
})(Slider);

class Envelope extends React.Component {
  constructor(props) {
    super(props);
    this.handleChange = this.handleChange.bind(this);
  }

  handleChange(event, value, type) {
    this.props.setEnvelope(type, value);
  }
  render() {
    return (
      <div className="envelope">
        <FormLabel component="legend">Envelope</FormLabel>
        <div className="graph">
          <EnvelopeGraph style={
            {height: "100%", width: "100%"}}
          a={this.props.envelope.attack}
          d={this.props.envelope.decay}
          s={this.props.envelope.sustain}
          r={this.props.envelope.release}
          />
        </div>
        <div className="sliders">

          <Typography id="label">A</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.attack}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "attack");
            }}
          />
          <Typography id="label">D</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.decay}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "decay");
            }}
          />
          <Typography id="label">S</Typography>
          <StyledSlider
            id="attack"
            className="slider"
            value={this.props.envelope.sustain}
            min={0}
            max={1}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "sustain");
            }}
          />
          <Typography id="label">R</Typography>
          <StyledSlider
            id="attack"
            className="slider release"
            value={this.props.envelope.release}
            min={0.01}
            max={2}
            aria-labelledby="label"
            onChange={(event, value) => {
              this.handleChange(event, value, "release");
            }}
          />
        </div>
      </div>
    );
  }
}

Envelope.propTypes = {
  setEnvelope: PropTypes.func.isRequired,
  envelope: PropTypes.objectOf(PropTypes.number).isRequired,
};

export default Envelope;
