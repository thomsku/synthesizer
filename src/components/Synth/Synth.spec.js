import React from "react";
import {shallow, mount} from "enzyme";
import Synth from "./Synth";
import AudioAnalyser from "../AudioAnalyser/AudioAnalyser";
import Tuning from "../Tuning/Tuning";
import Waveform from "../Waveform/Waveform";
import Envelope from "../Envelope/Envelope";
import {RadioGroup} from "@material-ui/core";
import {Slider} from "@material-ui/lab";
import EnvelopeGraph from "adsr-envelope-graph";
import {justKeyToFrequency, equalKeyToFrequency} from "./frequencies.js";
require("./AudioContext.mock.js");

describe("Synth", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Synth />));

  it("should render correctly", () => expect(wrapper).toMatchSnapshot());

  it("should render three <div />", () => {
    expect(wrapper.find("div")).toHaveLength(3);
  });

  it("should render the AudioAnalyser, Tuning, Waveform and Envelope Components", () => {
    expect(wrapper.containsAllMatchingElements([
      <AudioAnalyser
        key="AudioAnalyser"
        audioContext={wrapper.instance().audioContext}
        activeOscillators={wrapper.instance().state.activeOscillators}
        amp={wrapper.instance().volume}/>,
      <Tuning
        key="Tuning"
        setTuning={wrapper.instance().setTuning}
        tuning="just"/>,
      <Waveform
        key="Waveform"
        setWaveform={wrapper.instance().setWaveform}
        waveform="sine"/>,
      <Envelope
        key="Envelope"
        setEnvelope={wrapper.instance().setEnvelope}
        envelope={{attack: 0.1, decay: 0.01, sustain: 1, release: 0.1}}
      />,
    ])).toEqual(true);
  });
});

describe("mounted Synth", () => {
  let wrapper;
  window.HTMLCanvasElement.prototype.getContext = () => {
    return {
      width: 0,
      beginPath: jest.fn(),
      clearRect: jest.fn(),
      lineTo: jest.fn(),
      moveTo: jest.fn(),
      stroke: jest.fn(),
    };
  };

  beforeEach(() => wrapper = mount(<Synth />));

  it("calls play() when the right key is pressed", () => {
    const spy = jest.spyOn(wrapper.instance(), "play");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown({key: "a"});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("calls stop() when the right key is released", () => {
    const spy = jest.spyOn(wrapper.instance(), "stop");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown({key: "a"});
    wrapper.instance().handleKeyUp({key: "a"});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("does not call play() when the wrong key is pressed", () => {
    const spy = jest.spyOn(wrapper.instance(), "play");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown({key: "x"});
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it("does not call stop() when the wrong key is released", () => {
    const spy = jest.spyOn(wrapper.instance(), "stop");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.instance().handleKeyDown({key: "x"});
    wrapper.instance().handleKeyUp({key: "x"});
    expect(spy).toHaveBeenCalledTimes(0);
  });

  it("calls setTuning when the tuning radio buttons are changed", () => {
    const spy = jest.spyOn(wrapper.instance(), "setTuning");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(RadioGroup).at(0).find("input").at(1).simulate("change", {target: {value: "equal"}});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("calls setWaveform when the waveform radio buttons are changed", () => {
    const spy = jest.spyOn(wrapper.instance(), "setWaveform");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(RadioGroup).at(1).find("input").at(1).simulate("change", {target: {value: "square"}});
    expect(spy).toHaveBeenCalledTimes(1);
  });

  it("calls setEnvelope when the envelope sliders are changed", () => {
    const spy = jest.spyOn(wrapper.instance(), "setEnvelope");
    wrapper.instance().forceUpdate();
    expect(spy).toHaveBeenCalledTimes(0);
    wrapper.find(Slider).at(0).props().onChange(null, 0.2);
    expect(spy).toHaveBeenCalledTimes(1);
    wrapper.find(Slider).at(1).props().onChange(null, 0.2);
    expect(spy).toHaveBeenCalledTimes(2);
    wrapper.find(Slider).at(2).props().onChange(null, 0.2);
    expect(spy).toHaveBeenCalledTimes(3);
    wrapper.find(Slider).at(3).props().onChange(null, 0.2);
    expect(spy).toHaveBeenCalledTimes(4);
  });

  afterAll(() => wrapper.unmount());
});

describe("play", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Synth />));

  it("updates activeOscillators", () => {
    expect(wrapper.state("activeOscillators")).toEqual({});
    wrapper.instance().play("a");
    expect(wrapper.state("activeOscillators")).toHaveProperty("a");
  });

  it("plays an oscillator with the correct pitch", () => {
    wrapper.setState({tuning: {type: "equal", frequencies: equalKeyToFrequency}});
    wrapper.instance().play("d");
    expect(wrapper.state().activeOscillators["d"].oscillator.frequency.value).toEqual(329.63);
  });

  it("plays an oscillator with the correct waveform", () => {
    wrapper.setState({waveform: "triangle"});
    wrapper.instance().play("d");
    expect(wrapper.state().activeOscillators["d"].oscillator.type).toEqual("triangle");
  });

  it("plays an oscillator with the correct envelope", () => {
    const envelope = {attack: 1, decay: 1, sustain: 0.9, release: 1};
    wrapper.setState({envelope: envelope});
    wrapper.instance().play("d");
    expect(wrapper.state().activeOscillators["d"].envelope.attackTime).toEqual(envelope.attack);
    expect(wrapper.state().activeOscillators["d"].envelope.decayTime).toEqual(envelope.decay);
    expect(wrapper.state().activeOscillators["d"].envelope.sustainLevel).toEqual(envelope.sustain);
    expect(wrapper.state().activeOscillators["d"].envelope.releaseTime).toEqual(envelope.release);
  });
});

describe("setTuning", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Synth />));

  it("updates tuning type and frequency object in state", () => {
    expect(wrapper.state("tuning")).toEqual({type: "just", frequencies: justKeyToFrequency});
    wrapper.instance().setTuning({target: {value: "equal"}});
    expect(wrapper.state("tuning")).toEqual({type: "equal", frequencies: equalKeyToFrequency});
    wrapper.instance().setTuning({target: {value: "just"}});
    expect(wrapper.state("tuning")).toEqual({type: "just", frequencies: justKeyToFrequency});
  });
});

describe("setWaveform", () => {
  let wrapper;
  beforeEach(() => wrapper = shallow(<Synth />));

  it("updates waveform in state", () => {
    expect(wrapper.state("waveform")).toEqual("sine");
    wrapper.instance().setWaveform({target: {value: "square"}});
    expect(wrapper.state("waveform")).toEqual("square");
  });
});

describe("setEnvelope", () => {
  let wrapper;
  beforeEach(() => wrapper = mount(<Synth />));

  it("updates envelope in state", () => {
    expect(wrapper.state("envelope")).toEqual({attack: 0.1, decay: 0.01, sustain: 1, release: 0.1});
    wrapper.instance().setEnvelope("attack", 1);
    expect(wrapper.state("envelope")).toEqual({attack: 1, decay: 0.01, sustain: 1, release: 0.1});
  });

  it("updates EnvelopeGraph attack", () => {
    wrapper.instance().setEnvelope("attack", 1);
    wrapper.update();
    expect(wrapper.find(EnvelopeGraph).props().a).toEqual(1);
  });

  it("updates EnvelopeGraph decay", () => {
    wrapper.instance().setEnvelope("decay", 1);
    wrapper.update();
    expect(wrapper.find(EnvelopeGraph).props().d).toEqual(1);
  });

  it("updates EnvelopeGraph sustain", () => {
    wrapper.instance().setEnvelope("sustain", 0.9);
    wrapper.update();
    expect(wrapper.find(EnvelopeGraph).props().s).toEqual(0.9);
  });

  it("updates EnvelopeGraph release", () => {
    wrapper.instance().setEnvelope("release", 1);
    wrapper.update();
    expect(wrapper.find(EnvelopeGraph).props().r).toEqual(1);
  });
});
