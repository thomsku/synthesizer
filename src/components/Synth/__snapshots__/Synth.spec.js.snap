// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Synth should render correctly 1`] = `
<div
  className="synthDiv"
>
  <AudioAnalyser
    activeOscillators={Object {}}
    amp={
      GainNode {
        "gain": AudioGain {
          "value": 0.2,
        },
      }
    }
    audioContext={
      AudioContext {
        "destination": AudioDestinationNode {},
        "listener": AudioListener {},
      }
    }
  />
  <h2>
    Synth Parameters
  </h2>
  <div
    className="parameters"
  >
    <Tuning
      setTuning={[Function]}
      tuning="just"
    />
    <Waveform
      setWaveform={[Function]}
      waveform="sine"
    />
    <Envelope
      envelope={
        Object {
          "attack": 0.1,
          "decay": 0.01,
          "release": 0.1,
          "sustain": 1,
        }
      }
      setEnvelope={[Function]}
    />
  </div>
</div>
`;
